File: 4kyu/currying_vs_partial_application/solution_currying_vs_partial_application.py
"""https://www.codewars.com/kata/currying-vs-partial-application"""

from types import FunctionType


def curry_partial(f, *initial_args):
    """Curries and partially applies the initial arguments to the function."""
    if not isinstance(f, FunctionType):
        return f

    if f.__code__.co_argcount <= len(initial_args):
        return f(*(initial_args[:f.__code__.co_argcount] or initial_args))

    return lambda *args: curry_partial(f, *initial_args, *args)

File: javascript/src/solutions/4kyu/currying_vs_partial_application/solution_currying_vs_partial_application.js
/*
 * https://www.codewars.com/kata/53cf7e37e9876c35a60002c9
 */

export function curryPartial(fn, ...args) {
    if (fn.length <= args.length) {
        return fn(...args)
    }

    return (...x) => curryPartial(fn, ...args, ...x);
}
